src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift

```diff
@@ -0,0 +1,189 @@
+				.contains(where: { $0.name.standardizedName != healthCertificate.name.standardizedName })
+			if hasDifferentName {
+				return .failure(.nameMismatch)
+			}
+
+			let hasDifferentDateOfBirth = healthCertifiedPerson.healthCertificates
+				.contains(where: { $0.dateOfBirth != healthCertificate.dateOfBirth })
+			if hasDifferentDateOfBirth {
+				return .failure(.dateOfBirthMismatch)
+			}
+
+			healthCertifiedPerson.healthCertificates.append(healthCertificate)
+
+			if !healthCertifiedPersons.value.contains(healthCertifiedPerson) {
+				healthCertifiedPersons.value.append(healthCertifiedPerson)
+			}
+
+			if healthCertificate.isEligibleForProofCertificate {
+				healthCertifiedPerson.proofCertificateUpdatePending = true
+			}
+
+			return .success((healthCertifiedPerson))
+		} catch let error as CertificateDecodingError {
+			return .failure(.decodingError(error))
+		} catch {
+			return .failure(.other(error))
+		}
+	}
+
+	func removeHealthCertificate(_ healthCertificate: HealthCertificate) {
+		for healthCertifiedPerson in healthCertifiedPersons.value {
+			if let index = healthCertifiedPerson.healthCertificates.firstIndex(of: healthCertificate) {
+				healthCertifiedPerson.healthCertificates.remove(at: index)
+
+				if healthCertifiedPerson.healthCertificates.isEmpty {
+					healthCertifiedPersons.value.removeAll(where: { $0 == healthCertifiedPerson })
+				} else if healthCertificate.isEligibleForProofCertificate {
+					healthCertifiedPerson.proofCertificateUpdatePending = true
+				}
+
+				break
+			}
+		}
+	}
+
+	func updateProofCertificate(
+		for healthCertifiedPerson: HealthCertifiedPerson,
+		trigger: FetchProofCertificateTrigger,
+		completion: @escaping (Result<Void, HealthCertificateServiceError.ProofRequestError>) -> Void
+	) {
+		guard healthCertifiedPerson.shouldAutomaticallyUpdateProofCertificate || trigger != .automatic else {
+			Log.info("[HealthCertificateService] Not requesting proof for health certified person: \(private: healthCertifiedPerson). (proofCertificateUpdatePending: \(healthCertifiedPerson.proofCertificateUpdatePending), lastProofCertificateUpdate: \(String(describing: healthCertifiedPerson.lastProofCertificateUpdate)), trigger: \(trigger))", log: .api)
+
+			return
+		}
+
+		Log.info("[HealthCertificateService] Requesting proof for health certified person: \(private: healthCertifiedPerson). (proofCertificateUpdatePending: \(healthCertifiedPerson.proofCertificateUpdatePending), lastProofCertificateUpdate: \(String(describing: healthCertifiedPerson.lastProofCertificateUpdate)), trigger: \(trigger)", log: .api)
+
+		let healthCertificates = healthCertifiedPerson.healthCertificates
+			.filter { $0.isEligibleForProofCertificate }
+			.map { $0.base45 }
+
+		if healthCertificates.isEmpty {
+			healthCertifiedPerson.removeProofCertificateIfExpired()
+			completion(.success(()))
+		}
+
+		ProofCertificateDownload().fetchProofCertificate(
+			for: healthCertificates,
+			completion: { result in
```

Comment:
Not sure if completion ever will be called, if the ProofCertificateDownload instance is deallocated right away.